### ISSUE TITLE
TowerWeb future times out

### ISSUE THREAD

**Issue by gakonst** [tomusdrw/rust-web3#227]
I have an API built with https://github.com/carllerche/tower-web/. It works by enclosing a struct definition in a `impl_web!` macro and exposes api endpoints to any function you decorate.

Inside the API, I want to perform some ethereum-related action, for the example let's say fetching the tx count of an address. 

When using `.wait()` on the call, everything works fine. When chaining `and_then`'s such that this can work asynchronously, it errors out:

```
Io(
    Kind(
        TimedOut,
    ),
)
```

Interestingly enough, the RPC does not get hit at all! So the timeout error does not make much sense to me.

I've put a minimal example to reproduce and debug this issue. You must run `ganache-cli` or some ethereum node that has the rpc endpoint at localhost:8545
This might be related to tower_web, and I am investigating it at the moment. The TimedOut error occurs here: https://github.com/tomusdrw/rust-web3/blob/master/src/transports/shared.rs#L129-L130

https://github.com/gakonst/web3-rust-tower-web-repro

Output of `RUST_LOG=web3 cargo test`

```
[2019-07-08T16:19:37Z DEBUG web3::transports::http] [0] Sending: {"jsonrpc":"2.0","method":"eth_getTransactionCount","params":["0x3cdb3d9e1b74692bb1e3bb5fc81938151ca64b02","latest"],"id":0} to http://127.0.0.1:8545/
[2019-07-08T16:19:37Z DEBUG web3::transports::http] [0] Sending: {"jsonrpc":"2.0","method":"eth_getTransactionCount","params":["0x3cdb3d9e1b74692bb1e3bb5fc81938151ca64b02","latest"],"id":0} to http://127.0.0.1:8545/
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Request pending.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Checking response.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Request pending.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Checking response.
Couldn't fetch nonce. Got error: Io(
    Kind(
        TimedOut,
    ),
)
[2019-07-08T16:19:37Z DEBUG web3::transports::http] [0] Sending: {"jsonrpc":"2.0","method":"eth_getTransactionCount","params":["0x3cdb3d9e1b74692bb1e3bb5fc81938151ca64b02","latest"],"id":0} to http://127.0.0.1:8545/
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Request pending.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Checking response.
Couldn't fetch nonce. Got error: Io(
    Kind(
        TimedOut,
    ),
)
test tests::future_fails ... ok
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Checking response.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Extracting result.
[2019-07-08T16:19:37Z DEBUG web3::transports::http] [0] Sending: {"jsonrpc":"2.0","method":"eth_getTransactionCount","params":["0x3cdb3d9e1b74692bb1e3bb5fc81938151ca64b02","latest"],"id":0} to http://127.0.0.1:8545/
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Request pending.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Checking response.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Checking response.
[2019-07-08T16:19:37Z TRACE web3::transports::shared] [0] Extracting result.
test tests::wait_works ... ok
```

Thank you for putting this library together.

**Comment by tomusdrw** [comment]
Dropping the `_eloop` object in your examples kills the underlying `EventLoop` of `rust-web3`.

`rust-web3` is still using `tokio_core` instead of `tokio` (see #tokio), so it doesn't have a notion of "always-available-in-the-background-runtime", instead the event loop thread is started and controlled manually by the transport.

I just made a PR (see #228) to allow the `EventLoop` to run in the background, so you should  convert your example to:
```rust
        fn bar(&self) -> Box<dyn Future<Item = (), Error = ()> + Send> {
            let (eloop, transport) = Http::new(&self.endpoint).unwrap();
            // Run event loop in the background.
            eloop.into_remote();
            let web3 = Web3::new(transport);
```
and it should work fine.

**Comment by gakonst** [comment]
Amazing. Thank you so much @tomusdrw, works as expected now.

**Comment by gakonst** [comment]
I see, that makes sense. When using your solution, the `into_remote()` call hangs though. I've pushed to the repository to illustrate it. Tried with geth, parity and ganache-cli, hangs in the same place (tried websockets as well) https://github.com/gakonst/web3-rust-tower-web-repro/commit/66733c422b5cbaa30675049665605e31c1cf24d5

**Comment by tomusdrw** [comment]
Ugh, yeah sorry. I've implemented it incorrectly: see #230 for a fix.

**Comment by gakonst** [comment]
Fantastic. Just tested it and it works. Thank you.
